Start implementing layout engine

GabrielMajeri opened the pull request:
Starts work towards #6.

I've opened this to get some feedback.

Currently, it only computes sizes (not offsets), and it only works for the top-level window.

GabrielMajeri commented on src/layout_abstract_syntax.rs:
```diff
@@ -35,6 +35,16 @@ impl LayoutAbstractTag {
 		self.attributes.push(attribute);
 	}
+
+	/// Retrieves the `width` attribute of this node, if any.
+	pub fn width(&self) -> Option<&AttributeValue> {
+		self.attributes.iter().find(|attribute| attribute.name == "width").map(|attribute| &attribute.value)
+	}
+
+	/// Retrieves the `height` attribute of this node, if any.
+	pub fn height(&self) -> Option<&AttributeValue> {
+		self.attributes.iter().find(|attribute| attribute.name == "height").map(|attributes| &attributes.value)
+	}
```
I've created these helper functions, but it makes sense to eventually just store the layout attributes as a dedicated structure in the node (because these are some built-in attributes).

GabrielMajeri commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+}
+/// Built-in component attributes used during layout.
+#[derive(Debug)]
+struct LayoutAttributes {
+	pub width: Dimension,
+	pub height: Dimension,
+}
```
Eventually this will also store alignments and padding and such.

GabrielMajeri commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+	// Extract the layout-specific attributes of this node
+	let mut attribs = node.borrow().layout_attributes();
+
+	// For the sake of example, we'll assume the main window has a fixed size
+	if tag.namespace == "window" && tag.name == "main" {
+		attribs.width = Dimension::AbsolutePx(1920.0);
+		attribs.height = Dimension::AbsolutePx(1080.0);
+	}
```
This is also temporary. The `main.xml` file said the window would be called with a specific width and height, so for now I've taken some example values.

GabrielMajeri commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+		if let Dimension::AbsolutePx(size) = attribs.size_along(main_axis) {
+			size - min_main_axis_size
+		}
+		else {
+			panic!("child layout uses free space percentage but parent doesn't have a well-defined size");
+		}
```
This is a problematic scenario. What if the child wants to be 100@ of the free space of the parent, but the parent has an unbound 100% size?

The solution would be to always make sure the root of the hierarchy has a fixed size (for example, the window).

Keavon replied:
The window will indeed be always called with absolute pixel values, like shown on line 1 of `window/main.xml`, equal to the dimensions of the window frame.

Keavon commented on gui/README.md:
```diff
@@ -15,6 +15,23 @@ The layout engine does a top-down pass through the component tree in order to de
 
+### Layout algorithm
+
+To calculate dimensions (width/height):
+- elements with **fixed** sizes (e.g. `123px`) are laid out, in the order they are encountered.
+- then **percentages** (such as `10%`) are computed, based on the total size of the parent container.
+- then we handle elements with `inner`, which indicates that the size of the component depends on the size of the child.
+- then `@` rules (e.g. `100@`) are applied, which divide up the remaining free space.
```
`inner` elements will actually need to be calculated before `%` elements.

Think of `px` and `inner` as elements with outwards-pushing sizes. Their demanded size is set within (either the literal `px` value or by recursing into the `inner` element and solving for a pixel value). Then padding and spacing are added outside of them (padding grows the actual size, spacing just pushes the x/y alignment within its container and takes up remaining space).

After these two are calculated (`px` then `inner`), the container has been pushed to expand by these types of children. Its size is now known, so it can go push inwards on `%` and `@` elements.

Think of `%` and `@` as elements with sizes that are pushed inwards by their container. The container's size is now known, and these two types of elements have their dimensions based on that known size. `%` elements are calculated and that means all the dimensions are known except `@` elements. Add up all the known sizes and subtract that sum from the container's dimension to get the remaining unfilled pixels. Now the `@` elements are finally given a size equal to their percentage of the remainder. (`50@` = 50% of the remainder, so a `100px` remainder becomes an absolute `50px` size for that element.)

So in summary, the order goes like this:
1. Immediately count all the `px` values, these are done.
2. Recursively calculate `inner` sizes to get their pixel values, now all outwards-pushing elements are done.
3. If the container was `inner`, it now has a known size: `max(outwardsPusher + itsPadding + itsSpacing, ...)` pixels
4. Now `%` values are `percentage / 100 * containerSize` pixels
5. Finally `@` values are `atSignPercentage / 100 * (containerSize - (knownElementSizes + theirPadding + theirSpacing))` pixels

GabrielMajeri replied:
Thanks for the explanation, I've updated the `README.md` for GUI with a summary.

Keavon commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+/// An axis along which layout is performed
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Axis {
+	/// Horizontal axis, which is normally used by rows.
```
> normally used **within** rows

(rows themselves are stacked on top of one another if there are multiple sibling rows, and sibling `<row>`s and `<column>`s are an error by the way)

Not trying to nitpick but this is a distinction that could lead to confusion (I had to think about it for a few minutes).

GabrielMajeri replied:
I've reworded it, hopefully it makes more sense now.

Keavon commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+	/// Converts an `AttributeValue` to a `Dimension`.
+	fn from_attribute_value(value: &AttributeValue) -> Self {
+		if let AttributeValue::TypeValue(values) = value {
+			assert_eq!(values.len(), 1, "Unexpected number of values");
```
It might be more useful for this function to return a vector of `Dimension`s. It would make sense to treat everything as multi-value attributes (for example `padding="1px 2px 3px 4px"`), and single-value attributes are multi-value attributes of length 1.

GabrielMajeri replied:
I'll leave that as a to-do for the future

Keavon commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+
+				LayoutAttributes { width, height }
+			},
+			Self::Text(content) => {
+				// For now, we assume the text only takes up one row,
+				// and we have a monospaced font where each character is 20px wide.
+				let width = (content.len() * 20) as f32;
+				let height = 20.0;
+
+				let width = Dimension::AbsolutePx(width);
+				let height = Dimension::AbsolutePx(height);
+
+				LayoutAttributes { width, height }
+			},
```
We won't even need to bother with this branch, text nodes aren't handled by the layout engine. They exist so you can pass text to a particular node's `content` attribute. The `<text>` primitive node will be in charge of rendering text and since it is `inner` by default, it will usually report its width when the text renderer reports that. The text (if you bind it, like `content="SOME_VAR"`) can also be used in the JavaScript(?) scripts associated with each layout component. For example, the `<dropdown>` element will need a list of selectable dropdown items so the JS will be `SOME_VAR.split("\n").map(s => s.trim())`.

GabrielMajeri replied:
I've let it return width/height 100% for now so it doesn't crash my WIP cross-axis calculation code.

Keavon commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+			assert_eq!(values.len(), 1, "Unexpected number of values");
+
+			if let TypeValueOrArgument::TypeValue(ref value) = values[0] {
+				let dimension = match *value {
+					TypeValue::AbsolutePx(pixels) => Dimension::AbsolutePx(pixels),
+					TypeValue::Percent(percent) => Dimension::Percent(percent),
+					TypeValue::PercentRemainder(percent) => Dimension::PercentRemainder(percent),
+					TypeValue::Inner => Dimension::Inner,
+					TypeValue::Width => Dimension::Width,
+					TypeValue::Height => Dimension::Height,
+					_ => panic!("unsupported type in layout attribute"),
+				};
+				dimension
```
It might make sense to break this out into its own conversion function since it will be useful to call it in other places.

Keavon commented on src/layout_engine.rs:
```diff
@@ -0,0 +1,206 @@
+		attribs.height = Dimension::AbsolutePx(1080.0);
+	}
+
+	let main_axis = if tag.namespace == "" && tag.name == "row" { Axis::Horizontal } else { Axis::Vertical };
```
Okay, let's default to `Horizontal` (so things flow like they are inside a `<row>` if neither `<row>` nor `<col>` is specified. (`window/main.xml` is now being wrapped in `<col>`.)

GabrielMajeri replied:
Makes sense, horizontal is also the default for the web.

GabrielMajeri commented:
I've updated the PR. It is now able to recurse through the layout for `window:main` without crashing.

It seems to compute main axis sizes (i.e. heights) correctly. The cross axis sizes aren't really implemented yet.

GabrielMajeri commented:
I've updated this earlier, but it seems GitHub didn't display the changes :thinking: 

The changes are as follows:
- rebased on top of the latest `master`, fixing what changed.
- switched to using the `euclid` crate for 2D geometry. This is the same one used by the Servo browser's layout engine, so it's right on target.

GabrielMajeri commented:
I've rebased the PR, and incorporated the latest changes from `master`. The code now compiles and runs, and produces mostly the same output as before.

The code is still messy in my opinion and I'll look into breaking it up into multiple PRs or refactors over the following days.

Keavon commented:
Shelved on account of switching to a web app frontend for the short term.
